chain-tx-enclave-next/mls/src/tree_math.rs

```diff
@@ -28,125 +28,140 @@
 //!   So moving from parent to children or vice versa, is only about switching one or two bits.
 //!
 //! The left sub-tree of the root node is always complete, but the right part might not.
+use std::convert::{From, TryFrom};
+use std::iter;
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct LeafSize(pub u32);
 #[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
-pub struct LeafSize(pub usize);
+pub struct ParentSize(pub u32);
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct NodeSize(pub u32);
```

Comment:
I wonder if it's necessary to have the internal member as `pub`? there could instead be `Into` implementation -- or I guess this is needed in MLS types? Perhaps it could just be if these types have a decoder implementation?

Reply:
> should those two be self.ratchets[sender_leaf.node_index()] instead of self.ratchets[sender_leaf.0 as usize]?

`node_index()` returns the index in the binary tree(`leaf_i * 2`), but ratchets is indexed by `leaf_i` itself.

Reply:
> or Vec::with_capacity?

I think `resize_with` will do single heap allocation, so there should be no performance difference here.

Reply:
Ideally we should encapsulate tree math within a tree abstraction, together with the `Vec<Node>`. I'm not sure how far can we go that way without abstraction leak. I want to try this idea, also separating secrets from public tree in another PR.